File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListActivityRenderer.kt
package com.margelo.nitro.nativelist

import android.graphics.drawable.GradientDrawable
import android.text.TextUtils
import android.view.Gravity
import android.widget.LinearLayout
import com.facebook.react.uimanager.ThemedReactContext
import org.json.JSONObject

internal class NativeListActivityRowView(context: ThemedReactContext) :
  NativeListRendererRowView(context) {
  private val details = NativeListActivityDetailsView(context)
  private val content =
    LinearLayout(context).apply {
      orientation = HORIZONTAL
      gravity = Gravity.CENTER_VERTICAL
    }
  private val visual = NativeListLeadingVisual(context)
  private val column = LinearLayout(context).apply { orientation = VERTICAL }
  private val titleLine =
    LinearLayout(context).apply {
      orientation = HORIZONTAL
      gravity = Gravity.CENTER_VERTICAL
    }
  private val title = NativeListTextView(context)
  private val description = NativeListTextView(context)
  private val status = NativeListTextView(context)
  private val failure = NativeListTextView(context)
  private val amounts =
    LinearLayout(context).apply {
      orientation = VERTICAL
      gravity = Gravity.END
    }
  private val values = List(2) { NativeListTextView(context) }
  private val actions = LinearLayout(context).apply { orientation = HORIZONTAL }
  private val buttons = List(3) { NativeListTextView(context) }
  override val assetFields = listOf("leading", "secondaryLeading", "amounts")

  init {
    titleLine.addView(title, LayoutParams(0, LayoutParams.WRAP_CONTENT, 1f))
    titleLine.addView(failure)
    column.addView(titleLine)
    column.addView(description)
    column.addView(status)
    values.forEach(amounts::addView)
    buttons.forEach(actions::addView)
    content.addView(visual)
    content.addView(column, LayoutParams(0, LayoutParams.WRAP_CONTENT, 1f))
    content.addView(amounts)
    addView(content, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT))
    addView(actions)
    addView(details, LayoutParams(-1, -2))
    details.visibility = GONE
    details.onAction = { key, view, source, slot -> emitAction(key, view, source, slot) }
  }

  override fun defaultHeight(item: NativeListItem, layout: String) =
    if (item.json.has("amounts")) 0 else if ((item.json.optJSONArray("footerActions")?.length() ?: 0) > 0) 104 else 60

  override fun bindContent(
    item: NativeListItem,
    theme: JSONObject?,
    layout: String,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String,
  ) {
    val data = item.json
    val rich = data.has("amounts")
    details.visibility = if (rich) VISIBLE else GONE
    content.visibility = if (rich) GONE else VISIBLE
    actions.visibility = if (rich) GONE else VISIBLE
    if (rich) {
      orientation = VERTICAL
      val style = data.optJSONObject("style") ?: JSONObject()
      val hp = if (style.has("horizontalPadding")) stylePx(style.optDouble("horizontalPadding")) else dp(if (data.optString("presentation") == "table") 16 else 12)
      val vp = if (style.has("verticalPadding")) stylePx(style.optDouble("verticalPadding")) else dp(8)
      setPadding(hp, vp, hp, vp)
      visual.recycle()
      details.bind(item, theme, sourceScale)
      return
    }
    details.recycle()
    val style = data.optJSONObject("style") ?: JSONObject()
    val image = style.optJSONObject("image") ?: JSONObject()
    fun color(name: String, fallback: String) =
      parseNativeListColor(theme?.optString(name, fallback) ?: fallback)
    fun fill(value: Int, radius: Int) =
      GradientDrawable().apply {
        setColor(value)
        cornerRadius = dp(radius).toFloat()
      }
    fun bind(
      view: NativeListTextView,
      value: String,
      slot: String,
      size: Float,
      weight: String = "regular",
      ink: Int = color("secondaryText", "#0000009B"),
      line: Int = 0,
      lines: Int = 1,
      align: String = "start",
    ) {
      val textStyle =
        JSONObject(style.optJSONObject(slot)?.toString() ?: "{}").apply {
          if (!has("alignment")) put("alignment", align)
        }
      NativeListResolvedText.resolve(
          context,
          value,
          textStyle,
          size,
          weight,
          ink,
          line,
          lines,
          sourceScale,
        )
        .bind(view)
      view.ellipsize =
        if (textStyle.optString("truncate") == "clip") null else TextUtils.TruncateAt.END
    }
    orientation = VERTICAL
    val hasFooterActions = (data.optJSONArray("footerActions")?.length() ?: 0) > 0
    gravity =
      when (style.optJSONObject("container")?.optString("contentVerticalAlignment")) {
        "bottom" -> Gravity.BOTTOM
        "top" -> Gravity.TOP
        "center" -> Gravity.CENTER_VERTICAL
        // Legacy: the footer-action column is top/start aligned.
        else -> if (hasFooterActions) Gravity.START else Gravity.CENTER_VERTICAL
      }
    content.gravity =
      when (style.optJSONObject("container")?.optString("contentVerticalAlignment")) {
        "bottom" -> Gravity.BOTTOM
        "top" -> Gravity.TOP
        else -> Gravity.CENTER_VERTICAL
      }
    val hp =
      if (style.has("horizontalPadding")) stylePx(style.optDouble("horizontalPadding"))
      else dp(if (layout == "table") 16 else 12)
    val vp =
      if (style.has("verticalPadding")) stylePx(style.optDouble("verticalPadding")) else dp(8)
    setPadding(hp, vp, hp, vp)
    val iw = if (image.has("width")) stylePx(image.optDouble("width")) else dp(40)
    val ih = if (image.has("height")) stylePx(image.optDouble("height")) else dp(40)
    val gap = if (style.has("leadingGap")) stylePx(style.optDouble("leadingGap")) else dp(12)
    visual.layoutParams = LayoutParams(iw, ih).apply { marginEnd = gap }
    visual.bind(
      // Legacy: a missing visual still reserves an empty, transparent slot.
      data.optJSONObject("leading") ?: JSONObject().put("backgroundColor", "#00000000"),
      image,
      item.key,
      theme,
      false,
      sourceScale,
      data.optJSONObject("secondaryLeading"),
    )
    bind(
      title,
      data.optString("title"),
      "title",
      16f,
      "medium",
      color("primaryText", "#000000DF"),
      24,
    )
    bind(description, data.optString("description"), "description", 14f, line = 20, lines = 2)
    val failed = data.optString("status") == "Failed"
    bind(status, if (failed) "" else data.optString("status"), "status", 12f)
    bind(
      failure,
      if (failed) "Failed" else "",
      "status",
      12f,
      "medium",
      color("negative", "#C40006D3"),
    )
    failure.background = fill(color("criticalBackground", "#F3000D14"), 4)
    failure.setPadding(dp(8), dp(2), dp(8), dp(2))
    (failure.layoutParams as LayoutParams).apply {
      marginStart =
        if (style.has("titleBadgeGap")) stylePx(style.optDouble("titleBadgeGap")) else 0
    }
    for (view in listOf(description, status)) (view.layoutParams as LayoutParams).apply {
      topMargin = if (style.has("lineGap")) stylePx(style.optDouble("lineGap")) else 0
    }
    for ((index, slot) in listOf("primaryAmount", "secondaryAmount").withIndex()) {
      val value = data.optString(slot)
      bind(
        values[index],
        value,
        slot,
        if (index == 0) 16f else 14f,
        if (index == 0) "medium" else "regular",
        if (index == 0)
          color(
            if (value.startsWith("+")) "positive" else "primaryText",
            if (value.startsWith("+")) "#00713FDE" else "#000000DF",
          )
        // Legacy: both amounts default to primaryText; only "+" amounts are positive.
        else color("primaryText", "#000000DF"),
        line = if (index == 0) 24 else 20,
        align = "end",
      )
      values[index].layoutParams =
        LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
          if (index > 0)
            topMargin = if (style.has("trailingGap")) stylePx(style.optDouble("trailingGap")) else 0
        }
    }
    val descriptors = data.optJSONArray("footerActions")
    val count = minOf(3, descriptors?.length() ?: 0)
    actions.visibility = if (count == 0) GONE else VISIBLE
    actions.layoutParams =
      LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
        marginStart = if (style.has("leadingGap") || image.has("width")) iw + gap else dp(52)
        topMargin = if (style.has("lineGap")) stylePx(style.optDouble("lineGap")) else dp(8)
        bottomMargin = dp(4)
      }
    for ((index, button) in buttons.withIndex()) {
      button.visibility = if (index < count) VISIBLE else GONE
      button.setOnClickListener(null)
      if (index >= count) continue
      val action = descriptors!!.getJSONObject(index)
      NativeListResolvedText.resolve(
          context,
          action.optString("label"),
          null,
          // Legacy footer action: 14sp semibold on a 20dp line.
          14f,
          "semibold",
          color(
            if (action.optString("tone") == "danger") "negative" else "primaryText",
            if (action.optString("tone") == "danger") "#C40006D3" else "#000000DF",
          ),
          20,
          1,
          sourceScale,
        )
        .bind(button)
      button.background = fill(color("strongBackground", "#0000000F"), 8)
      button.setPadding(dp(8), dp(4), dp(8), dp(4))
      button.isEnabled = !data.optBoolean("disabled") && !action.optBoolean("disabled")
      button.layoutParams =
        LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
          marginEnd = dp(8)
        }
      button.setOnClickListener {
        emitAction(action.optString("key"), button, "footerAction", index)
      }
    }
  }

  override fun recycleContent() {
    visual.recycle()
    details.recycle()
    buttons.forEach { it.setOnClickListener(null) }
  }

  override fun disposeContent() {
    visual.dispose()
    details.dispose()
    buttons.forEach { it.setOnClickListener(null) }
  }
}

File: native-views/react-native-native-list/ios/NativeListActivityRenderer.swift
import UIKit

final class NativeListActivityCell: NativeListRendererCell {
  private let visual = NativeListLeadingVisual()
  private let column = UIStackView()
  private let titleLine = UIStackView()
  private let title = NativeListTextLabel()
  private let descriptionLabel = NativeListTextLabel()
  private let status = NativeListTextLabel()
  private let failure = NativeListTextLabel()
  private let amounts = UIStackView()
  private let amountLabels = [
    NativeListAccessoryButton(type: .system), NativeListAccessoryButton(type: .system),
  ]
  private let actions = UIStackView()
  private let buttons = (0..<3).map { _ in UIButton(type: .system) }
  private var actionKeys: [String] = []
  private var dimensions: [NSLayoutConstraint] = []
  override var assetFields: [String] { ["leading", "secondaryLeading"] }

  override init(frame: CGRect) {
    super.init(frame: frame)
    root.axis = .horizontal
    column.axis = .vertical
    column.alignment = .fill
    column.setContentHuggingPriority(UILayoutPriority(1), for: .horizontal)
    column.setContentCompressionResistancePriority(.defaultHigh, for: .horizontal)
    titleLine.axis = .horizontal
    titleLine.alignment = .center
    titleLine.spacing = 8
    title.setContentCompressionResistancePriority(.required, for: .horizontal)
    failure.setContentHuggingPriority(.required, for: .horizontal)
    titleLine.addArrangedSubview(title)
    titleLine.addArrangedSubview(failure)
    [titleLine, descriptionLabel, status, actions].forEach(column.addArrangedSubview)
    amounts.axis = .vertical
    amounts.alignment = .trailing
    amounts.setContentHuggingPriority(.required, for: .horizontal)
    amounts.setContentCompressionResistancePriority(.required, for: .horizontal)
    amountLabels.forEach(amounts.addArrangedSubview)
    actions.axis = .horizontal
    actions.alignment = .center
    actions.spacing = 8
    for (index, button) in buttons.enumerated() {
      button.tag = index
      button.titleLabel?.font = nativeListFont(ofSize: 12, weight: .medium)
      button.layer.cornerRadius = 8
      button.contentEdgeInsets = UIEdgeInsets(top: 4, left: 8, bottom: 4, right: 8)
      button.addTarget(self, action: #selector(pressed(_:)), for: .touchUpInside)
      actions.addArrangedSubview(button)
    }
    [visual, column, amounts].forEach(root.addArrangedSubview)
  }
  required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
  override func bindContent(
    _ item: NativeListItem, theme: [String: Any]?, layout: String,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String
  ) {
    let data = item.data
    let style = data.dictionary("style") ?? [:]
    let image = style.dictionary("image") ?? [:]
    root.spacing = 12
    root.setCustomSpacing(CGFloat(style.double("leadingGap", default: 12)), after: visual)
    switch style.dictionary("container")?.string("contentVerticalAlignment") {
    case "top": root.alignment = .top
    case "bottom": root.alignment = .bottom
    default: root.alignment = .center
    }
    column.spacing = CGFloat(style.double("lineGap"))
    titleLine.spacing = CGFloat(style.double("titleBadgeGap", default: 8))
    amounts.spacing = CGFloat(style.double("trailingGap", default: 2))
    contentInsets = UIEdgeInsets(
      top: CGFloat(style.double("verticalPadding", default: 8)),
      left: CGFloat(style.double("horizontalPadding", default: layout == "table" ? 16 : 12)),
      bottom: CGFloat(style.double("verticalPadding", default: 8)),
      right: CGFloat(style.double("horizontalPadding", default: layout == "table" ? 16 : 12)))
    NSLayoutConstraint.deactivate(dimensions)
    dimensions = [
      visual.widthAnchor.constraint(equalToConstant: CGFloat(image.double("width", default: 40))),
      visual.heightAnchor.constraint(equalToConstant: CGFloat(image.double("height", default: 40))),
    ]
    NSLayoutConstraint.activate(dimensions)
    visual.bind(
      data.dictionary("leading") ?? [:], style: image, key: item.key, theme: theme, isUnread: false,
      secondaryVisual: data.dictionary("secondaryLeading"))
    func bind(
      _ view: NativeListTextLabel, _ value: String, _ slot: String, size: CGFloat,
      weight: NativeListFontWeight = .regular, color: UIColor, line: CGFloat, lines: Int = 1
    ) {
      NativeListResolvedText(
        value, style: style.dictionary(slot), size: size, weight: weight, color: color,
        lineHeight: line, lines: lines
      ).bind(view)
    }
    bind(
      title, data.string("title"), "title", size: 16, weight: .medium,
      color: nativeListColor(theme, "primaryText", "#202020"), line: 24)
    bind(
      descriptionLabel, data.string("description"), "description", size: 14,
      color: nativeListColor(theme, "secondaryText", "#646464"), line: 20, lines: 2)
    let failed = data.string("status") == "Failed"
    bind(
      status, failed ? "" : data.string("status"), "status", size: 12,
      color: nativeListColor(theme, "secondaryText", "#646464"),
      line: nativeListFont(ofSize: 12).lineHeight)
    bind(
      failure, failed ? "  Failed  " : "", "status", size: 12, weight: .medium,
      color: nativeListColor(theme, "negative", "#CE2C31"),
      line: nativeListFont(ofSize: 12, weight: .medium).lineHeight)
    failure.backgroundColor = nativeListColor(theme, "criticalBackground", "#F3000D14")
    failure.layer.cornerRadius = 4
    failure.clipsToBounds = true
    for (index, slot) in ["primaryAmount", "secondaryAmount"].enumerated() {
      let button = amountLabels[index]
      let value = data.string(slot)
      let color =
        index == 0
        ? nativeListColor(
          theme, value.hasPrefix("+") ? "positive" : "primaryText",
          value.hasPrefix("+") ? "#218358" : "#202020")
        : nativeListColor(theme, "secondaryText", "#646464")
      button.titleLabel?.numberOfLines = 1
      button.titleLabel?.lineBreakMode = .byTruncatingTail
      button.contentHorizontalAlignment = .center
      button.contentVerticalAlignment = .center
      button.rowTextOffsetY = 0
      let resolved = NativeListResolvedText(
        value, style: nil, size: index == 0 ? 16 : 14, weight: index == 0 ? .medium : .regular,
        color: color, lineHeight: index == 0 ? 24 : 20, lines: 1, tabular: true)
      button.setAttributedTitle(resolved.attributed, for: .normal)
      button.titleLabel?.font = resolved.font
      button.isHidden = value.isEmpty
      if let text = style.dictionary(slot) { NativeListTextStyles.applyStyledButton(button, text) }
    }
    let descriptors = Array(data.dictionaries("footerActions").prefix(3))
    actions.isHidden = descriptors.isEmpty
    actionKeys = descriptors.map { $0.string("key") }
    for (index, button) in buttons.enumerated() {
      button.isHidden = index >= descriptors.count
      guard index < descriptors.count else { continue }
      let action = descriptors[index]
      button.setTitle(action.string("label"), for: .normal)
      button.isEnabled = !data.bool("disabled") && !action.bool("disabled")
      button.backgroundColor = nativeListColor(theme, "strongBackground", "#F0F0F0")
      button.setTitleColor(
        nativeListColor(
          theme, action.string("tone") == "danger" ? "negative" : "primaryText",
          action.string("tone") == "danger" ? "#CE2C31" : "#202020"), for: .normal)
    }
  }
  @objc private func pressed(_ button: UIButton) {
    guard button.tag < actionKeys.count else { return }
    emitAction(actionKeys[button.tag], from: button, source: "footerAction", slot: button.tag)
  }
  override func recycleContent() {
    visual.recycle()
    actionKeys = []
    buttons.forEach { $0.isHidden = true }
  }
}
